### PR TITLE
Int and float toString() produces numeric-string

### PR DESCRIPTION
**Pull request by b1rdex**
This is for https://github.com/phpstan/phpstan/issues/4005

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
Hi, this needs a regression test. Probably add a new dataProvider in NodeScopeResolverTest::testFileAsserts() and test that `(string) $int` and `(string) $float` return `string&numeric`.

**Comment by b1rdex** [comment]
@ondrejmirtes done

File: src/Type/Constant/ConstantIntegerType.php [hunk 1]
```diff
@@ -74,7 +76,10 @@ public function toFloat(): Type
 
 	public function toString(): Type
 	{
-		return new ConstantStringType((string) $this->value);
+		return new IntersectionType([
+			new ConstantStringType((string) $this->value),
+			new AccessoryNumericStringType(),
```
Comment (ondrejmirtes) [inline src/Type/Constant/ConstantIntegerType.php:81]: This isn't necessary, `ConstantStringType` will already say that it's numeric based on the value.

File: tests/PHPStan/Analyser/data/cast-to-numeric-string.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ */
+function foo(int $a, float $b, $numeric, $numeric2, $number, $positive, $negative, $constantInt): void {
+	assertType('string&numeric', (string)$a);
+	assertType('string&numeric', (string)$b);
+	assertType('string&numeric', (string)$numeric);
+	assertType('string&numeric', (string)$numeric2);
+	assertType('string&numeric', (string)$number);
+	assertType('string&numeric', (string)$positive);
+	assertType('string&numeric', (string)$negative);
+	assertType('\'1\'&numeric', (string)$constantInt);
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/cast-to-numeric-string.php:23]: This should be just `'1'`.

File: src/Type/Accessory/AccessoryNumericStringType.php [hunk 3]
```diff
@@ -138,7 +138,10 @@ public function toFloat(): Type
 
 	public function toString(): Type
 	{
-		return $this;
+		return new IntersectionType([
+			new StringType(),
+			$this,
+		]);
```
Comment (ondrejmirtes) [inline src/Type/Accessory/AccessoryNumericStringType.php:144]: This isn't necessary IMHO.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by b1rdex** [comment]
@ondrejmirtes fixed

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**